### ISSUE TITLE
FB13580167: AVAudioRecorder is missing support for voice isolation

### ISSUE THREAD

**Issue by sindresorhus** [feedback-assistant/reports#465]
- **Date:** 2024-02-03
- **Resolution:** Open
- **Area:** AVFoundation
- **OS:** iOS 17.3
- **Type:** Incorrect/Unexpected Behavior

## Description

[Voice isolation](https://support.apple.com/en-us/HT213683#:~:text=During%20your%20call%2C%20swipe%20down,Voice%20Isolation) is a very useful feature on iOS to improve voice recording. However, it is currently not supported when recording audio using AVAudioRecorder. AVAudioRecorder is the easiest and most popular way to record audio for apps on iOS, so it means that most apps will not benefit from this feature, which is sad. To support voice isolation, apps needs to rewrite their audio recording using AVAudioEngine, which is much more complicated and will likely cause bugs to be introduced.

**Comment by m13v** [comment]
yeah this is frustrating, voice isolation works great in FaceTime/Phone but Apple hasn't exposed it through AVAudioRecorder or AVAudioEngine. the workaround we use is AVAudioEngine with the voice processing audio unit (kAudioUnitSubType_VoiceProcessingIO) which gives you echo cancellation and noise suppression, though it's not exactly the same as the system-level voice isolation.

for better results, we ended up using a post-processing approach - record the raw audio and then run a noise reduction model on it. the RNNoise-based approach works surprisingly well for voice isolation and runs in real-time on Apple Silicon. not as seamless as the system feature but it works with any audio source, not just calls.

**Comment by m13v** [comment]
we handle audio recording with noise processing in our voice-first macOS agent. the audio pipeline with AVAudioEngine and voice processing: https://github.com/m13v/fazm/blob/main/Desktop/Sources/TranscriptionService.swift

and the screen capture service that records alongside audio: https://github.com/m13v/macos-session-replay/blob/main/Sources/SessionReplay/ScreenCaptureService.swift
